packages/vite/src/node/utils.ts

```diff
@@ -38,8 +38,9 @@ export const normalizeId = (id: string): string =>
   id.replace(/(\s*>\s*)/g, ' > ')
 
 export function isBuiltin(id: string): boolean {
-  const deepMatch = id.match(deepImportRE)
-  id = deepMatch ? deepMatch[1] || deepMatch[2] : id
+  const builtins = [...builtinModules]
+  if (!builtins.includes('fs/promises')) builtins.push('fs/promises')
```

Comment:
I prefer to maintain a hard-coded list like @patak-js [said](https://github.com/vitejs/vite/pull/5809#issuecomment-977957201). Here only fs/promises are dealt with specially, other modules will still be affected by the version of node. For example, v14 has the `diagnostics_channel` module but v12 does not.

Reply:
I like the idea of avoiding reducing what we are harcoding, but if it isn't only `fs/promises` then it looks safer to have the full list. It would be harder to debug if there are differences in resolution depending on the node version used. If new builtins are added, we would still have to add them by hand to support v12 anyways so we don't gain in maintenance burden of keeping it up to date.

Reply:
added `diagnostics_channel`

Reply:
Just an example😂, the complete list of differences I manually checked is as follows: 

```
['assert/strict','diagnostics_channel','dns/promises','fs/promises','path/posix','path/win32' ,'readline/promises','stream/consumers','stream/promises','stream/web','timers/promises','util/types','wasi']
```

I also like your idea of reducing hard coding, but at least we have to make sure that all possible differences are listed.

Reply:
true. I won't be home from work before late evening

Reply:
I agree too to maintain a separate hardcoded list for all builtin modules. Here we are already maintaining a partial list of modules between node 12 & 16, we could go all in and do a full list.

Reply:
I prefer this way without too much hardcoding to keep vite package size as small as possible

Reply:
Could generate the hard-coded list at build time

Reply:
But why do we bother recognizing the newest built-in modules in older Node.js versions?
Say, if the user uses `fs/promises' in Node.js v12, there *should* be an error.

Reply:
@sodatea Because we need to know whether we are parsing a built-in module or a local module file? If `isBuiltin` mistakenly judges the built-in module as a local module file, and `nodeResolve` should be used but `viteResolve` is used, it will cause unexpected errors.

Reply:
@ygj6 But `fs/promises` is not a built-in module in Node.js 12. So if the user is in Node.js 12, an error is expected.

Reply:
@sodatea 
>an error is expected.

I agree with you. The problem is that the error message thrown may mislead users, thinking that this is an internal error of vite, but did not realize that the node version used does not have this module?

For example, the correct one is to throw an error like this:
```
Error: Cannot find module'fs/promises' from'/Users/me/project/vite-ssr-http-common-issue'
......
```
but accidentally throws an error like this:
```
Error: Failed to resolve import "_http_common" from "ssr-module.mjs". Does the file exist?
......
```

Is this acceptable, or is there another way to avoid it?

Reply:
But to me the second one looks more clearer. And it's not wrong…

Reply:
If it would be only for the error message, I agree with @sodatea here.

I was under the impression that this was needed due to SSR builds where we could be compiling using node v12 code to be run in node v16 in the server. I know it is an edge case, but I don't see anything preventing this. Here https://github.com/vitejs/vite/blob/650b56e0414d772294817bc2dc399e70563ed699/packages/vite/src/node/plugins/resolve.ts#L243
But maybe I misunderstood the issue.

Reply:
Yeah I can second that this is mainly for better error messages only. Besides SSR, for prebundling, we would externalize node builtins with [a custom proxy](https://github.com/vitejs/vite/blob/650b56e0414d772294817bc2dc399e70563ed699/packages/vite/src/node/optimizer/esbuildDepPlugin.ts#L193-L206) to provide better error messages. I'm not sure what the error message would look like if we don't externalize it, but browser ESM often gives generic and hard-to-debug errors.